File: epi_judge_cpp/adding_credits.cc
#include <string>
#include <unordered_map>
#include <unordered_set>

#include "test_framework/fmt_print.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
using std::string;

class ClientsCreditsInfo {
 public:
  void Insert(const string& client_id, int c) {
      Remove(client_id);
      client_to_credit_.emplace(client_id, c - offset_);
      credit_to_client_[c - offset_].emplace(client_id);
  }

  bool Remove(const string& client_id) {
      auto iter_client = client_to_credit_.find(client_id);
      if (iter_client != client_to_credit_.end()) {
          credit_to_client_[iter_client->second].erase(client_id);
          if (credit_to_client_[iter_client->second].empty()) {
              credit_to_client_.erase(iter_client->second);
          }

          client_to_credit_.erase(iter_client);
          return true;
      }

      return false;
  }
  int Lookup(const string& client_id) const {
      auto iter = client_to_credit_.find(client_id);
      return (iter == client_to_credit_.end()) ? -1 : iter->second + offset_;
  }

  void AddAll(int C) {
      offset_ += C;
  }
  
  string Max() const {
      auto iter = credit_to_client_.crbegin();
      return (iter == credit_to_client_.crend() || iter->second.empty()) ? "" : *iter->second.cbegin();
 
  }

private:
    int offset_ = 0;
    std::unordered_map<string, int> client_to_credit_;
    std::map<int, std::unordered_set<string>> credit_to_client_;

};
struct Operation {
  std::string op;
  std::string s_arg;
  int i_arg;
};

std::ostream& operator<<(std::ostream& out, const Operation& op) {
  return out << FmtStr("{}({}, {})", op.op, op.s_arg, op.i_arg);
}

namespace test_framework {
template <>
struct SerializationTrait<Operation>
    : UserSerTrait<Operation, std::string, std::string, int> {};
}  // namespace test_framework
void ClientsCreditsInfoTester(const std::vector<Operation>& ops) {
  ClientsCreditsInfo cr;
  int op_idx = 0;
  for (auto& op : ops) {
    if (op.op == "ClientsCreditsInfo") {
      continue;
    } else if (op.op == "remove") {
      bool result = cr.Remove(op.s_arg);
      if (result != op.i_arg) {
        throw TestFailure()
            .WithProperty(PropertyName::STATE, cr)
            .WithProperty(PropertyName::COMMAND, op)
            .WithMismatchInfo(op_idx, op.i_arg, result);
      }
    } else if (op.op == "max") {
      auto result = cr.Max();
      if (result != op.s_arg) {
        throw TestFailure()
            .WithProperty(PropertyName::STATE, cr)
            .WithProperty(PropertyName::COMMAND, op)
            .WithMismatchInfo(op_idx, op.i_arg, result);
      }
    } else if (op.op == "insert") {
      cr.Insert(op.s_arg, op.i_arg);
    } else if (op.op == "add_all") {
      cr.AddAll(op.i_arg);
    } else if (op.op == "lookup") {
      auto result = cr.Lookup(op.s_arg);
      if (result != op.i_arg) {
        throw TestFailure()
            .WithProperty(PropertyName::STATE, cr)
            .WithProperty(PropertyName::COMMAND, op)
            .WithMismatchInfo(op_idx, op.i_arg, result);
        ;
      }
    }
    op_idx++;
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"ops"};
  return GenericTestMain(args, "adding_credits.cc", "adding_credits.tsv",
                         &ClientsCreditsInfoTester, DefaultComparator{},
                         param_names);
}

File: epi_judge_java_solutions/epi/AddingCredits.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.EpiUserType;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.Set;
import java.util.TreeMap;

public class AddingCredits {

  public static class ClientsCreditsInfo {

    private int offset = 0;
    private Map<String, Integer> clientToCredit = new HashMap<>();
    private NavigableMap<Integer, Set<String>> creditToClients =
        new TreeMap<>();

    public void insert(String clientID, int c) {

      remove(clientID);
      clientToCredit.put(clientID, c - offset);
      creditToClients.putIfAbsent(c - offset, new HashSet<>());
      Set<String> set = creditToClients.get(c - offset);
      set.add(clientID);
    }

    public boolean remove(String clientID) {

      Integer clientCredit = clientToCredit.get(clientID);
      if (clientCredit != null) {
        creditToClients.get(clientCredit).remove(clientID);
        if (creditToClients.get(clientCredit).isEmpty()) {
          creditToClients.remove(clientCredit);
        }
        clientToCredit.remove(clientID);
        return true;
      }
      return false;
    }

    public int lookup(String clientID) {

      Integer clientCredit = clientToCredit.get(clientID);
      return clientCredit == null ? -1 : clientCredit + offset;
    }

    public void addAll(int C) { offset += C; }

    public String max() {

      return creditToClients.isEmpty()
          ? ""
          : creditToClients.lastEntry().getValue().iterator().next();
    }

    @Override
    public String toString() {

      return "{clientToCredit=" + clientToCredit + '}';
    }
  }

  @EpiUserType(ctorParams = {String.class, String.class, int.class})
  public static class Operation {
    public String op;
    public String sArg;
    public int iArg;

    public Operation(String op, String sArg, int iArg) {
      this.op = op;
      this.sArg = sArg;
      this.iArg = iArg;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }

      Operation operation = (Operation)o;

      if (iArg != operation.iArg) {
        return false;
      }
      if (!op.equals(operation.op)) {
        return false;
      }
      return sArg.equals(operation.sArg);
    }

    @Override
    public int hashCode() {
      int result = op.hashCode();
      result = 31 * result + sArg.hashCode();
      result = 31 * result + iArg;
      return result;
    }

    @Override
    public String toString() {
      return String.format("%s(%s, %d)", op, sArg, iArg);
    }
  }

  @EpiTest(testDataFile = "adding_credits.tsv")
  public static void ClientsCreditsInfoTester(List<Operation> ops)
      throws TestFailure {
    ClientsCreditsInfo cr = new ClientsCreditsInfo();
    int opIdx = 0;
    for (Operation x : ops) {
      String sArg = x.sArg;
      int iArg = x.iArg;
      int result;
      switch (x.op) {
      case "ClientsCreditsInfo":
        break;
      case "remove":
        result = cr.remove(sArg) ? 1 : 0;
        if (result != iArg) {
          throw new TestFailure()
              .withProperty(TestFailure.PropertyName.STATE, cr)
              .withProperty(TestFailure.PropertyName.COMMAND, x)
              .withMismatchInfo(opIdx, iArg, result);
        }
        break;
      case "insert":
        cr.insert(sArg, iArg);
        break;
      case "add_all":
        cr.addAll(iArg);
        break;
      case "lookup":
        result = cr.lookup(sArg);
        if (result != iArg) {
          throw new TestFailure()
              .withProperty(TestFailure.PropertyName.STATE, cr)
              .withProperty(TestFailure.PropertyName.COMMAND, x)
              .withMismatchInfo(opIdx, iArg, result);
        }
      }
      opIdx++;
    }
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "AddingCredits.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
